proxy.go

```diff
@@ -218,9 +220,14 @@ func (p *Proxy) Serve(l net.Listener) error {
 }
 
 func (p *Proxy) handleLoop(conn net.Conn) {
+	p.connsMu.Lock()
 	p.conns.Add(1)
+	p.connsMu.Unlock()
 	defer p.conns.Done()
 	defer conn.Close()
+	if p.Closing() {
```

Comment:
This is technically an independent change, but seems like a good idea. I can remove it if you prefer.